s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -232,6 +240,128 @@ import scala.util.{Failure, Success, Try}
       .mapMaterializedValue(_.flatMap(identity)(ExecutionContexts.parasitic))
   }
 
+  def getObjectByRanges(
+                         s3Location: S3Location,
+                         versionId: Option[String],
+                         s3Headers: S3Headers,
+                         rangeSize: Long = DefaultByteRangeSize,
+                         parallelism: Int = 4
+                       ): Source[ByteString, Future[ObjectMetadata]] = {
+    Source.fromMaterializer { (_, _) =>
+      val objectMetadataMat = Promise[ObjectMetadata]()
+      getObjectMetadata(s3Location.bucket, s3Location.key, versionId, s3Headers)
+        .flatMapConcat {
+          case Some(s3Meta) if s3Meta.contentLength == 0 =>
+            objectMetadataMat.success(s3Meta)
+            Source.empty[ByteString]
+          case Some(s3Meta) =>
+            objectMetadataMat.success(s3Meta)
+            doGetByRanges(s3Location, versionId, s3Headers, s3Meta.contentLength, rangeSize, parallelism)
+          case None =>
+            val exc = new NoSuchElementException(s"Object does not exist at location [${s3Location.mkString}]")
+            objectMetadataMat.failure(exc)
+            Source.failed(exc)
```

Comment:
This will close the downstream ASAP, do you want to defer it?

Reply:
Hmmm not sure of what that implies ?
The idea is that no `ObjectMetadata` means no S3 object, so I think the source should fail as well as the materialized Future